### PR TITLE
fixed looping in bookmark title query execution. added proximity query

### PR DESCRIPTION
**Pull request by ymatsuda**
- fixed 100% load issue (hopefully)
  - there is no direct evidence, but it is most likely that the bug in LineQuery.emptyQueryNode is the cause.
  - emptyQueryNode is used when translation from a Lucene query is failed. From the log, there was a query that causes translation failure around the time the problem started.
- my first proximity query implementation
  - NOT enabled yet.
  - I will integrate it tomorrow

### CODE REVIEW

File: shoebox/app/com/keepit/search/query/ProximityQuery.scala [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+import org.apache.lucene.search.DocIdSetIterator
+import org.apache.lucene.util.PriorityQueue
+import org.apache.lucene.util.ToStringUtils
+import scala.collection.JavaConversions._
+import scala.math._
+
+object ProximityQuery {
+  def apply(fieldName: String, query: Query): ProximityQuery = {
+    val terms: Set[Term] = {
+      val termSet: java.util.Set[Term] = new java.util.HashSet[Term]
```
Comment (eishay) [inline shoebox/app/com/keepit/search/query/ProximityQuery.scala:22]: a common way to manage these fully qualified names is using

``` scala
import java.util.{HashSet => JHashSet, Set => JSet}
...
val termSet: JSet[Term] = new JHashSet[Term]
```

File: shoebox/app/com/keepit/search/query/ProximityQuery.scala [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+      termSet.toSet.filter{ _.field() == fieldName }
+    }
+    
+    apply(terms)
+  }
+  
+  def apply(terms: Set[Term]): ProximityQuery = new ProximityQuery(terms)
+  
+  val scoreFactorHalfDecay = 5
+  val scoreFactorTable = {
```
Comment (eishay) [inline shoebox/app/com/keepit/search/query/ProximityQuery.scala:33]: is that for optimization? cool!

File: shoebox/app/com/keepit/search/query/ProximityQuery.scala [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+      while (top.doc == doc && top.pos < Int.MaxValue) {
+        top.nextPos()
+        top = pq.updateTop()
+        if (prev eq top) {
+          prev = top
+          prevPos = top.pos
+          prevWeight = top.weight
+        } else {
+          if (prevPos < top.pos) {
+            sum += (prevWeight + top.weight) * ProximityQuery.scoreFactor(top.pos - prevPos)
```
Comment (eishay) [inline shoebox/app/com/keepit/search/query/ProximityQuery.scala:202]: nice!

**Comment by eishay** [comment]
@ymatsuda 
we're getting the following warning, can they be removed?

```
info] Compiling 8 Scala sources and 5 Java sources to /home/fortytwo/code/keepit/shoebox/target/scala-2.9.1/classes...
[warn] /home/fortytwo/code/keepit/shoebox/app/com/keepit/search/query/ProximityQuery.scala:47: class Searcher in package search is deprecated: see corresponding Javadoc for more information.
[warn]   override def createWeight(searcher: Searcher): Weight = new ProximityWeight(this, searcher.getSimilarity)
[warn]                                       ^
[warn] one warning found
[info] Compiling 57 Scala sources and 6 Java sources to /home/fortytwo/code/keepit/shoebox/target/scala-2.9.1/classes...
[warn] /home/fortytwo/code/keepit/shoebox/app/com/keepit/search/query/ProximityQuery.scala:47: class Searcher in package search is deprecated: see corresponding Javadoc for more information.
[warn]   override def createWeight(searcher: Searcher): Weight = new ProximityWeight(this, searcher.getSimilarity)
[warn]    
```

**Comment by eishay** [comment]
deployed in prod, note that when i deployed the server was on 100% again

File: shoebox/app/com/keepit/search/query/ProximityQuery.scala (hunk 1 above)
Comment (ymatsuda) [inline shoebox/app/com/keepit/search/query/ProximityQuery.scala:22]: Cool. That looks nicer.

File: shoebox/app/com/keepit/search/query/ProximityQuery.scala (hunk 2 above)
Comment (ymatsuda) [inline shoebox/app/com/keepit/search/query/ProximityQuery.scala:33]: yup

**Comment by ymatsuda** [comment]
Regarding the compiler warning, Lucene is using its own deprecated class in the interface definition. Do you like to suppress the warning?

**Comment by ymatsuda** [comment]
I could not find an annotation for suppressing deprecation warning.

**Pull request merged**